x-pack/platform/plugins/shared/cases/public/client/helpers/get_observables_from_ecs.ts

```diff
@@ -0,0 +1,162 @@
+    {
+      defaultMessage: 'Auto extracted observable',
+    }
+  );
+
+  // Source IP
+  if (ecsData.source?.ip) {
+    const ips = Array.isArray(ecsData.source?.ip) ? ecsData.source?.ip : [ecsData.source?.ip];
+    ips.forEach((ip) => {
+      const ipType = getIPType(ip);
+      observables.push({
+        typeKey: ipType === 'IPV4' ? 'observable-type-ipv4' : 'observable-type-ipv6',
+        value: ip,
+        description,
+      });
+    });
+  }
+
+  // Destination IP
+  if (ecsData.destination?.ip) {
+    const ips = Array.isArray(ecsData.destination.ip)
+      ? ecsData.destination.ip
+      : [ecsData.destination.ip];
+
+    ips.forEach((ip) => {
+      const ipType = getIPType(ip);
+      observables.push({
+        typeKey: ipType === 'IPV4' ? 'observable-type-ipv4' : 'observable-type-ipv6',
+        value: ip,
+        description,
+      });
+    });
+  }
+
+  // Host name
+  if (ecsData.host?.name) {
+    const hostnames = Array.isArray(ecsData.host?.name) ? ecsData.host?.name : [ecsData.host?.name];
+    hostnames.forEach((name) => {
+      observables.push({
+        typeKey: 'observable-type-hostname',
+        value: name,
+        description,
+      });
+    });
+  }
+
+  // File hash
+  const hashValues = getHashValues(ecsData);
+  if (hashValues.length > 0) {
+    observables.push(
+      ...hashValues.map((hash) => ({
+        typeKey: 'observable-type-file-hash',
+        value: hash,
+        description,
+      }))
+    );
+  }
+
+  // File path
+  if (ecsData.file?.path) {
+    const paths = Array.isArray(ecsData.file?.path) ? ecsData.file?.path : [ecsData.file?.path];
+    paths.forEach((path) => {
+      observables.push({
+        typeKey: 'observable-type-file-path',
+        value: path,
+        description,
+      });
+    });
+  }
+
+  // Domain;
+  if (ecsData.dns?.question?.name) {
+    const names = Array.isArray(ecsData.dns?.question?.name)
+      ? ecsData.dns?.question?.name
+      : [ecsData.dns?.question?.name];
+    names.forEach((name) => {
+      observables.push({
+        typeKey: 'observable-type-domain',
+        value: name,
+        description,
+      });
+    });
+  }
+
+  // URL
+  // TODO - Pending review
+  // https://github.com/elastic/security-team/issues/13709
+
+  // TODO - Pending review
+  // https://github.com/elastic/security-team/issues/13709
+  // email.from.address, or email.sender.address??
+  // if (ecsData.email?.from?.address) {
+  //   const addresses = Array.isArray(ecsData.email?.from?.address)
+  //     ? ecsData.email?.from?.address
+  //     : [ecsData.email?.from?.address];
+  //   observables.push(
+  //     ...addresses.map((address) => ({
+  //       typeKey: 'observable-type-email',
+  //       value: address,
+  //       description,
+  //     }))
+  //   );
+  // }
+
+  // remove duplicates of key type and value pairs
+  return [...new Set(observables.map((obj) => JSON.stringify(obj)))].map((str) => JSON.parse(str));
```

Comment:
Nifty, but 2 quick ideas

1. could shorten this as `[...new Set(Observables.map(JSON.stringify))].map(JSON.parse);`,
2. More importantly, could be better from a performance standpoint and safer to just keep a running ledger of what's been added, so something like

`addedObservables = new Map<typeof OBSERVABLE_TYPES_BUILTIN_KEYS, unknown[]>`. 

I don't think you have a union on the keys, so you may just simplify that as a string, but the type safety would be a nice addition. Then you can just add the observable key and value as you go on and only add it to the observables list if it's not already set

Reply:
yep, this is a very good call. I'll do the same on server side as well